path_metadata.go

```diff
@@ -136,11 +145,71 @@ func (b *versionedKVBackend) pathMetadataRead() framework.OperationFunc {
 				"max_versions":         meta.MaxVersions,
 				"cas_required":         meta.CasRequired,
 				"delete_version_after": deleteVersionAfter.String(),
+				"custom_metadata":      meta.CustomMetadata,
 			},
 		}, nil
 	}
 }
 
+const maxCustomMetadataKeys = 64
+const maxCustomMetadataKeyLength = 128
+const maxCustomMetadataValueLength = 512
+const customMetadataValidationErrorPrefix = "custom_metadata validation failed"
+
+// Perform input validation on custom_metadata field. If the key count
+// exceeds maxCustomMetadataKeys, the validation will be short-circuited
+// to prevent unnecessary (and potentially costly) validation to be run.
+// If the key count falls at or below maxCustomMetadataKeys, multiple
+// checks will be made per key and value. These checks include:
+//   - 0 < length of key <= maxCustomMetadataKeyLength
+//   - 0 < length of value <= maxCustomMetadataValueLength
+//   - keys and values cannot include unprintable characters
+func validateCustomMetadata(customMetadata map[string]string) error {
+	var errs *multierror.Error
+
+	if keyCount := len(customMetadata); keyCount > maxCustomMetadataKeys {
+		errs = multierror.Append(errs, fmt.Errorf("%s: payload must contain at most %d keys, provided %d",
+			customMetadataValidationErrorPrefix,
+			maxCustomMetadataKeys,
+			keyCount))
+
+		return errs.ErrorOrNil()
+	}
+
+	// Perform validation on each key and value and return ALL errors
+	for key, value := range customMetadata {
+		if keyLen := len(key); 0 == keyLen || keyLen > maxCustomMetadataKeyLength {
+			errs = multierror.Append(errs, fmt.Errorf("%s: length of key %q is %d but must be 0 < len(key) <= %d",
+				customMetadataValidationErrorPrefix,
+				key,
+				keyLen,
+				maxCustomMetadataKeyLength))
+		}
+
+		if valueLen := len(value); 0 == valueLen || valueLen > maxCustomMetadataValueLength {
+			errs = multierror.Append(errs, fmt.Errorf("%s: length of value for key %q is %d but must be 0 < len(value) <= %d",
+				customMetadataValidationErrorPrefix,
+				key,
+				valueLen,
+				maxCustomMetadataValueLength))
+		}
+
+		if !strutil.Printable(key) {
+			errs = multierror.Append(errs, fmt.Errorf("%s: key %q contains unprintable characters",
```

Comment:
any chance of an issue with having unprintable characters in the error?

Reply:
I see your test hits that case - so probably fine?

Reply:
A recent change to use `%q` in the format string will produce a quoted string including the unprintable character. It might make sense, however, to include both the string with the unprintable character exposed and without (i.e. format with `%s`) so that the user can easily identity the input. What do you think?

Reply:
yeah, I think that's a good idea!

Reply:
Great, just committed a change for this.